### PR TITLE
Batch update

### PR DESCRIPTION
**Pull request by EmteZogaf**
I simply tried to fix any unit test error. I still get an NPE occasionally in [DatabaseAsyncTest#testPushEmptyList](https://github.com/davidmoten/rxjava-jdbc/blob/batch-updates-3/src/test/java/com/github/davidmoten/rx/jdbc/DatabaseTestBase.java#L255-L269), but it's not happening while debugging. Maybe a race condition?

```
java.lang.RuntimeException: java.sql.SQLException: failed to execute sql=update person set score = score + 1
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:455)
	at rx.observables.BlockingObservable.single(BlockingObservable.java:331)
	at com.github.davidmoten.rx.jdbc.DatabaseTestBase.assertIs(DatabaseTestBase.java:959)
	at com.github.davidmoten.rx.jdbc.DatabaseTestBase.testPushEmptyList(DatabaseTestBase.java:268)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: java.sql.SQLException: failed to execute sql=update person set score = score + 1
	at com.github.davidmoten.rx.jdbc.QueryUpdateOnSubscribe.performUpdate(QueryUpdateOnSubscribe.java:236)
	at com.github.davidmoten.rx.jdbc.QueryUpdateOnSubscribe.call(QueryUpdateOnSubscribe.java:86)
	at com.github.davidmoten.rx.jdbc.QueryUpdateOnSubscribe.call(QueryUpdateOnSubscribe.java:1)
	at rx.Observable.unsafeSubscribe(Observable.java:7710)
	at rx.internal.operators.OperatorSubscribeOn$1$1.call(OperatorSubscribeOn.java:62)
	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.h2.jdbc.JdbcSQLException: General error: "java.lang.NullPointerException"
General error: "java.lang.NullPointerException" [50000-178]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:344)
	at org.h2.message.DbException.get(DbException.java:167)
	at org.h2.message.DbException.convert(DbException.java:294)
	at org.h2.message.DbException.toSQLException(DbException.java:267)
	at org.h2.message.TraceObject.logAndConvert(TraceObject.java:372)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:149)
	at com.github.davidmoten.rx.jdbc.PreparedStatementBatch.executeUpdate(PreparedStatementBatch.java:96)
	at com.github.davidmoten.rx.jdbc.QueryUpdateOnSubscribe.performUpdate(QueryUpdateOnSubscribe.java:220)
	... 12 more
Caused by: java.lang.NullPointerException
	at org.h2.jdbc.JdbcStatement.setExecutingStatement(JdbcStatement.java:1032)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:160)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:144)
	... 14 more

```

### CODE REVIEW

**Comment by EmteZogaf** [comment]
Tests are all green now. I rebased on current master. I'll try it in my application to see if it's actually working.

**Comment by EmteZogaf** [comment]
Rebased on current master (commit 2041df4)

**Comment by EmteZogaf** [comment]
@davidmoten had you any chance to look at the code? I have some more time now to work on this feature, but I wanted to know your thoughts first.

**Comment by davidmoten** [comment]
Just looking at your changes now, thank you very much for chasing this!

**Comment by thomasnield** [comment]
What's the status on this? Do conflicts need to be resolved or is this requiring a sizable amount of review?

**Comment by EmteZogaf** [comment]
I rebased on current master.

@davidmoten You're probably busy but any chance for some feedback?

**Comment by davidmoten** [comment]
Apologies guys, this will need a days work that I have to find. A bunch of unit tests need to be written just to ensure things like resource disposal for starters and to touch the usual edge cases.  I would like to do it in the next two weeks because I have a need for it at work as well. Feel free to bug me about it, I do want to get this in.

**Comment by thomasnield** [comment]
Okay awesome. I don't have the deep knowledge on all the implementation of this library, but please let me know if there is anything I can do to help.

File: src/main/java/com/github/davidmoten/rx/jdbc/ConnectionBatch.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    @Override
+    public boolean getAutoCommit() throws SQLException {
+        return con.getAutoCommit();
+    }
+
+    @Override
+    public void commit() throws SQLException {
+        if (Batch.get().enabled() && Batch.get().countAdded() > 0) {
+            Batch.get().getPreparedStatement().executeBatch();
+            Batch.set(Batch.get().reset());
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/ConnectionBatch.java]: This can just be `Batch.get().reset()`?

File: src/main/java/com/github/davidmoten/rx/jdbc/ConnectionBatch.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            throws SQLException {
+        return wrap(con.prepareStatement(sql, autoGeneratedKeys), sql, autoGeneratedKeys);
+    }
+
+    private static PreparedStatement wrap(PreparedStatement ps, String sql, int keysOption) throws SQLException {
+        if (Batch.get().getPreparedStatement() == null
+                || Batch.get().getPreparedStatement().getConnection() != ps.getConnection()
+                || !Batch.get().getPreparedStatement().getSql().equals(sql)
+                || Batch.get().getPreparedStatement().getKeysOption() != keysOption) {
+            Batch.set(Batch.get().setPreparedStatement(new PreparedStatementBatch(ps, sql, keysOption)));
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/ConnectionBatch.java]: Don't need the `Batch.set`?

File: src/main/java/com/github/davidmoten/rx/jdbc/PreparedStatementBatch.java [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+    @Override
+    public boolean isWrapperFor(Class<?> iface) throws SQLException {
+        return ps.isWrapperFor(iface);
+    }
+
+    @Override
+    public int executeUpdate(String sql) throws SQLException {
+        Batch batch = Batch.get();
+        if (batch.enabled()) {
+            Batch.set(batch.setPreparedStatement(this));
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/PreparedStatementBatch.java]: Don't need the `Batch.set`?

File: src/main/java/com/github/davidmoten/rx/jdbc/PreparedStatementBatch.java [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+
+    @Override
+    public void setCharacterStream(int parameterIndex, Reader reader, int length)
+            throws SQLException {
+        ps.setCharacterStream(parameterIndex, reader, length);
+    }
+
+    @Override
+    public void clearBatch() throws SQLException {
+        Batch.set(Batch.get().reset());
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/PreparedStatementBatch.java]: Don't need the `Batch.set`?

File: src/main/java/com/github/davidmoten/rx/jdbc/State.java [hunk 5]
```diff
@@ -6,8 +6,8 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 
 class State {
-    volatile Connection con;
-    volatile PreparedStatement ps;
-    volatile ResultSet rs;
+    Connection con;
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/State.java]: Don't think you can remove the volatile on these because unsubscription can happen from another thread and they must be visible.

File: src/main/java/com/github/davidmoten/rx/jdbc/ConnectionProviderSingletonManualCommitStatementCaching.java [hunk 6]
```diff
@@ -20,7 +20,7 @@
     /**
      * Ensures thread-safe setting of con
      */
-    private AtomicBoolean connectionSet = new AtomicBoolean(false);
+    private final AtomicBoolean connectionSet = new AtomicBoolean(false);
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rx/jdbc/ConnectionProviderSingletonManualCommitStatementCaching.java]: Thanks!

**Comment by davidmoten** [comment]
I've rebased batch-updates on master (should make easier to review).

**Comment by davidmoten** [comment]
What approach are you taking with generated keys resultset when batch is enabled? Are we assuming it's not supported?

**Comment by EmteZogaf** [comment]
Thanks for review and rebasing. I'm on vacation till friday next week, but I will look into your comments in detail right when I'm back.

**Comment by davidmoten** [comment]
Thanks @EmteZogaf

**Comment by EmteZogaf** [comment]
@davidmoten The current implementation does not allow returning generated keys combined with batch execution (see [QueryUpdate#returnGeneratedKeys](https://github.com/davidmoten/rxjava-jdbc/blob/batch-updates/src/main/java/com/github/davidmoten/rx/jdbc/QueryUpdate.java#L320-L324)). 

Returning generated keys from batch inserts depends on the jdbc implementation. A quick research revealed this:

**Support**:
- MySQL supports it. The order of the returned generated keys is guaranteed only of InnoDB tables with innodb_autoinc_lock_mode set to 0 (see Stackoverflow: [_Using getGeneratedKeys with batch..._](http://stackoverflow.com/questions/4952316/using-getgeneratedkeys-with-batch-inserts-in-mysql-with-connector-j#answer-4964797) and [_How can I Insert many rows into a MySQL table..._](http://stackoverflow.com/questions/7333524/how-can-i-insert-many-rows-into-a-mysql-table-and-return-the-new-ids#answer-16592867)).
- MariaDB supports it since version 1.1.8 (see [MariaDB Bug #83](https://mariadb.atlassian.net/browse/CONJ-83?jql=project%20%3D%20CONJ%20AND%20fixVersion%20%3D%201.1.8)).
- HSQLDB supports it (tested with in-memory db).
- DB2 does support it with the limitation that you can't combine it with an atomic multirow insert (see [DB2 docs](https://www.ibm.com/support/knowledgecenter/SSEPGG_9.5.0/com.ibm.db2.luw.apdv.java.doc/src/tpc/imjcc_c0057055.html)).

**Parital Support**:
- PostgreSQL supports it but gives no guarantee about the order of the returned keys (see [pgjdbc pr#204](https://github.com/pgjdbc/pgjdbc/pull/204)).
- H2 DB only returns the last key (see [H2 DB docs](http://h2database.com/javadoc/org/h2/jdbc/JdbcStatement.html#getGeneratedKeys)).

**No Support**:
- Oracle DB 12 does not support it (see [oracle docs](http://docs.oracle.com/database/121/JJDBC/jdbcvers.htm#JJDBC28104) in section _Limitations_).
- MS SQL Server does not seem to support it although there isn't any explicit statement (see [Stackoverflow: _getGeneratedKeys() after PreparedStatement.executeBatch()_](http://stackoverflow.com/questions/13641832/getgeneratedkeys-after-preparedstatement-executebatch)).


I think, we should support it and enhance the notes in the docs about the lacking support in some jdbc implementations. But I think we should leave this feature to be subject of another PR later on as this PR is intended to just fix unit test failures.

File: src/main/java/com/github/davidmoten/rx/jdbc/State.java (hunk 5 above)
Comment (EmteZogaf) [inline src/main/java/com/github/davidmoten/rx/jdbc/State.java]: You're probably right. I would like to have an unit test for this case, but couldn't get one ready after a quick hacking. I will try to finish this later next week.

**Comment by thomasnield** [comment]
I should add SQLite does not seem to support it either.

**Comment by EmteZogaf** [comment]
@thomasnield Thanks, added to list (found a source on Stackoverflow).

**Comment by EmteZogaf** [comment]
@davidmoten reverted back to volatile properties. You're right, unsubscribing in `QuerySelectOnSubscribe` could potentially happen in another thread.

**Comment by thomasnield** [comment]
What's the status on this?

**Comment by davidmoten** [comment]
Thanks @EmteZogaf for addressing the comments (ages ago, disculpa). What do you think the status of this PR is? 

@thomasnield  I'm guessing that our code coverage will be low and that some more tests need to be written before we have confidence in this one.  Once we have ok coverage I'll make an RC release and off we go.

**Comment by thomasnield** [comment]
@davidmoten do you guys need help writing tests?

**Comment by thomasnield** [comment]
What do you need tested? And on what database platforms?

**Comment by davidmoten** [comment]
@thomasnield yep, help greatly appreciated. I'll be driven by coverage to start with. I'll merge commits from @EmteZogaf into this branch now.

**Comment by EmteZogaf** [comment]
@davidmoten thx for merging. I'll think about more test cases in the next days. The next step is to guarantee the returning of generated keys for database which support them.

**Comment by thomasnield** [comment]
@EmteZogaf I think that is more on JDBC implementation end regarding correct return of generated keys. I found it to work correctly on the platforms that supported it. But I guess it wouldn't hurt to inventory what does and does not return generated keys.

**Comment by thomasnield** [comment]
Cannot replicate this test failure locally. It's working fine so I'll figure out what nuances are happening with Travis CI later. Regarding coverage, I am seeing you added `Batch`, `ConnectionBatch`, and `PreparedStatementBatch` types. Did you want tests  created for these types? Or just want to ensure they are exhaustively functional as whole in usage?
